Update intro and plotting (#8)
duke-malaria-collaboratory/DiscovR: _episodes/01-r-getting-started.md

> exercises: 15

Reviewer: Check to see if the times here match the schedule.

> 
> ~~~
> # this is a comment
> ~~~
> {: .language-r}
{: .callout}

# Foundational topics

## Functions

Functions are built-in procedures that automate a task for you.
You input arguments into a function and the function returns a value. 
We'll go over a few math functions to get our feet wet.

You call a function in R by typing it's name followed by opening then closing parenthesis. Each function has a purpose, which is often hinted at by the name of the function. 

Let's start with the `sqrt()` function.

Let's try to run the function without anything inside the parenthesis.

~~~
sqrt()
~~~
{: .language-r}

~~~
Error in sqrt(): 0 arguments passed to 'sqrt' which requires 1
~~~
{: .error}

We get an error message. Don't panic! Error messages pop up all the time, and can be super helpful in debugging code.

In this case, the message tells us zero arguments were passed to the function, but we need to input at least one. 
Many functions, including `sqrt`, require additional pieces of information to do their job. 
We call these additional values "arguments" or "parameters." 
You pass **arguments** to a function by placing values in between the parenthesis. 

> A function takes in these arguments and does a bunch of "magic" behind the scenes to output something we're interested in.

Reviewer: Instead of "magic," maybe we can say: A function takes in  these arguments and works behind the scenes to output something we're interested in.

Author: ```suggestion
A function takes in these arguments and works behind the scenes to output something we're interested in.
```

For example, we want to provide a number to `sqrt()`, namely the number we want the square root of: 

~~~
sqrt(4)
~~~
{: .language-r}

~~~
[1] 2
~~~
{: .output}

Here, the input argument is 4, and the output is 2, just like we'd expect. 

Now let's do an example where we might not know the expected output:

~~~
sqrt(2)
~~~
{: .language-r}

~~~
[1] 1.414214
~~~
{: .output}

Great, now let's move onto a slightly more complicated function. 
If we want to round a number, we can use the `round()` function:

~~~
round(3.14159)
~~~
{: .language-r}

~~~
[1] 3
~~~
{: .output}

Why did this round to three? What if we want it to round to a different number of digits?

> ## Pro-tip
>
> Each function has a help page that documents what arguments the function
> expects and what value it will return. You can bring up the help page a few
> different ways. If you have typed the function name in the **Editor** windows,
> you can put your cursor on the function name and press <kbd>F1</kbd> to open
> help page in the **Help** viewer in the lower right corner of RStudio. You can
> also type `?` followed by the function name in the console.
>

> > For example, try running `?round`. A help page should pop up with

Reviewer: We can be more specific and say: For example, try running `?round` in the console.

> information about what the function is used for and how to use it, as well as
> useful examples of the function in action. As you can see, 
`round()` has two arguments: the numeric input and the number of digits to round to.
>
{: .callout}

We can use the digits argument in `round()` to change how many decimal places are kept:

~~~
round(3.14159, digits = 2)
~~~
{: .language-r}

~~~
[1] 3.14
~~~
{: .output}

Sometimes it is helpful - or even necessary - to include the argument name, but often we can skip the argument name, if the argument values are passed in the order they are defined:

~~~
round(3.14159, 2)
~~~
{: .language-r}

~~~
[1] 3.14
~~~
{: .output}

> ## Position of the arguments in functions
> Which of the following lines of code will give you an output of 3.14? For the one(s) that don't give you 3.14, what do they give you?
>
> 
> ~~~
> round(x = 3.1415)

> > round("3.14159", 2)

Reviewer: We haven't introduced strings yet. Is this the best place to do it?

> round(x = 3.1415, digits = 2)
> round(digits = 2, x = 3.1415)
> round(2, 3.1415)
> ~~~
> {: .language-r}
> {: .source}
>
> > ## Solution

> > > The 2nd and 3rd lines will give you the right answer because the arguments are named, and when you use names the order doesn't matter. The 1st line will give you 3 because the default number of digits is 0. Then 4th line will give you 2 because, since you didn't name the arguments, x=2 and digits=3.1415.

Reviewer: I think these need revision:
- The 3rd and 4th lines will give you the right answer...
- The 1st line will give you 3...
- The 2nd line will give you an error because we've provided a non-numeric argument (string)
- The 5th line will give you 2 because...

Since we have them listed like this, would it be easier to just label them (A), (B), (C), etc?

> {: .solution}
{: .challenge}

> ## Bonus Exercise: taking logarithms
> Calculate the following:
> 1. Natural log (ln) of 10

> > 1. Log base 10 of 10 (extra points if you can do this in 2 different ways), and 

Reviewer: Instead of extra points, should we just say (challenge: try to do this 2 different ways)
